internal/resourcecache/resourcecache.go

```diff
@@ -0,0 +1,109 @@
+package resourcecache
+
+import (
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+var sleepTimeBeforeCleanup = 2 * time.Minute
+
+// ResourceCache is a structure that saves information about a nearly created resource.
+// Resources can be added and retrieved from the cache. A retrieval (ResourceFromCache) also removes the item from the cache.
+// If `sleepTimeBeforeCleanup` has passed, and the resource has not yet been retrieved, ResourceCache considers that resource stale,
+// and removes it.
+type ResourceCache struct {
+	resources map[string]*Resource
+	sync.Mutex
+}
+
+// Resource contains the actual resource itself (which must implement the IdentifiableCreatable interface),
+// as well as stores function pointers that pertain to how that resource should be cleaned up.
+type Resource struct {
+	resource     IdentifiableCreatable
+	cleanupFuncs []func()
+}
+
+// IdentifiableCreatable are the qualities needed by the caller of the resource.
+// Once a resource is retrieved, SetCreated() will be called, indicating to the server
+// that resource is ready to be listed and operated upon, and ID() will be used to identify the
+// newly created resource to the server.
+type IdentifiableCreatable interface {
+	ID() string
+	SetCreated()
+}
+
+// New creates a new ResourceCache
+func New() *ResourceCache {
+	rc := &ResourceCache{
+		resources: make(map[string]*Resource),
+	}
+	return rc
+}
+
+// AddResourceToCache takes a unique resource name (retrieved from the client request, not generated by the server)
+// a newly created resource, and functions to cleanup that newly created resource.
+// It adds the Resource to the ResourceCache, as well as starts a go routine that is responsible for cleaning up the
+// resource if the server has not gotten another request for it (i.e. if ResourceFromCache has not been called).
+// AddResourceToCache expects `name`s passed to it to be unique. If a duplicate `name` is detected, it returns an error.
+func (rc *ResourceCache) AddResourceToCache(name string, resource IdentifiableCreatable, cleanupFuncs []func()) error {
```

Comment:
`s/AddResourceToCache/Add/`?

Reply:
On a second thought, maybe it's better to have `Put` (instead of `AddResourceToCache`) and `Get` (instead of `ResourceFromCache`), and also `Got` (instead of `SetCreated`).

Reply:
`SetCreated` is a function of Container and Sandbox that marks it as created, so I don't as much want to change that, though ACK on the other changes